Plug-Ins/NovaTools/System Editor/SystemMapView.swift

```diff
@@ -0,0 +1,307 @@
+    }
+
+    func restackStellars() {
+        // Keep track of occupied locations - only the first stellar at a given point will be displayed
+        // This logic is inherited from GalaxyView where it's required for replacing systems based on visibilty;
+        // it's used here for consistency and just in case there's a reason to do the same with stellars.
+        var topViews: [NSPoint: StellarView] = [:]
+        selectedStellars = []
+        for view in controller.stellarViews.values {
+            view.highlightCount = 1
+            if let topView = topViews[view.point] {
+                if !view.isHighlighted {
+                    view.isHidden = true
+                } else if !topView.isHighlighted {
+                    // The current top view is not part of the selection, swap it with this one
+                    topView.isHidden = true
+                    topViews[view.point] = view
+                    view.isHidden = false
+                } else if topView != view {
+                    topView.highlightCount += 1
+                    view.isHidden = true
+                }
+            } else {
+                // No current top view, set it to this one
+                topViews[view.point] = view
+                view.isHidden = false
+            }
+            if view.isHighlighted {
+                selectedStellars.append(view)
+            }
+        }
+    }
+
+    // MARK: - Drawing
+
+    override var needsDisplay: Bool {
+        get { layer?.needsDisplay() ?? false }
+        set { layer?.setNeedsDisplay() }
+    }
+
+    override func makeBackingLayer() -> CALayer {
+        let layer = CALayer()
+        layer.contentsScale = window?.backingScaleFactor ?? 1
+        layer.backgroundColor = .black
+        layer.delegate = self
+        return layer
+    }
+
+    nonisolated func layer(_ layer: CALayer, shouldInheritContentsScale newScale: CGFloat, from window: NSWindow) -> Bool {
+        return true
+    }
+
+    func draw(_ layer: CALayer, in ctx: CGContext) {
+        NSGraphicsContext.current = NSGraphicsContext(cgContext: ctx, flipped: true)
+
+        // Centre line
+        let path = NSBezierPath()
+        NSColor(deviceRed: 0.5, green: 0.5, blue: 0.5, alpha: 0.4).setStroke()
+        path.move(to: NSPoint(x: 0, y: frame.size.height / 2 - 0.5))
+        path.line(to: NSPoint(x: frame.size.width, y: frame.size.height / 2 - 0.5))
+        path.move(to: NSPoint(x: frame.size.width / 2 - 0.5, y: 0))
+        path.line(to: NSPoint(x: frame.size.width / 2 - 0.5, y: frame.size.height))
+        path.stroke()
+    }
+
+    // MARK: - Pan and zoom
+
+    // Rescale all zoom levels so the 5k x 5k view can show the whole system (30k x 30k) even on max zoom
+    private let zoomLevels: [Double] = [8/19, 9/16, 3/4, 1, 4/3, 16/10]
+    func scaleFactor(forLevel level: Int) -> CGFloat {
+        return zoomLevels[level] * (5 / 30); // Scale the view to no more than 5k pixels across to avoid issues
+    }
```

Comment:
I removed the 1x option because it wasn't working with my previous approach to scaling the view (you couldn't actually see the whole system at 1x); now that I'm also setting the frame size (overriding the value configured in the xib) it works OK except once the size exceeds 10k the view can no longer be redrawn — the stellar subviews are positioned and shown correctly but the X/Y axis lines disappear. It's a bit weird but probably an acceptable tradeoff considering the scroll bar position provides some sense of where you are in the system, and you can navigate by clicking the nav defaults anyway.

Reply:
Ohh, I overlooked the part where you were changing the frame size. I see what you mean now.
Playing around with it again though, I still think it's important to provide a good experience in the system centre where people will be putting 90% of their spobs. So I think it would be best if we keep the view fixed at 8k to enable correct drawing at 1x, while still allowing people to zoom out to 1/4 (or further?) to access the edges when they need to. I don't think we need to worry too much that the edges won't be accessible at 1x.

Reply:
So I was fiddling with this and tried doing the above but... I was wrong. It doesn't work well if your wormhole disappears out of reach when you zoom in 😂
I realised drawing performance of the main view isn't a concern since it's pretty much static so the CALayer isn't really needed. When I replaced this with a standard draw routine it worked fine regardless of size. I also refactored the zooming a bit to fix an issue where the view wasn't initially centered correctly when you open it.